packages/rocketchat-channel-settings/server/models/Rooms.coffee

```diff
@@ -7,3 +7,42 @@ RocketChat.models.Rooms.setDescriptionById = (_id, description) ->
 			description: description
 
 	return @update query, update
+
+RocketChat.models.Rooms.setReadOnlyById = (_id, readOnly) ->
+	query =
+		_id: _id
+
+	update =
+		$set:
+			ro: readOnly
+
+	if readOnly
+		# we want to mute all users without the post-read-only permission
+
+		# get all usernames for this room
+		users = @findOne(query, { fields: { usernames: 1 }})
+
+		users.usernames.forEach (userName) ->
+
+			# lookup the user
+			user = RocketChat.models.Users.findOneByUsername userName
```

Comment:
This seems quite expensive everytime. Can you just request the required fields e.g. `{fields: {user._id: 1}}`?

Alternatively it might be more efficient to request all the user ids at once e.g.

```coffeescript
usernames = @findOne(query, { fields: { usernames: 1 }})
users = RocketChat.models.Users.findUsersByUsernames usernames, {fields: {username: 1}}
users.forEach (user) ->
     #Use user._id and user.username
```